### ISSUE TITLE
Question: data-flow promises in axios

### ISSUE THREAD

**Issue by D19cannon** [axios/axios#1898]
Hi! I had a quick question about the data flow of the response.

So, basically what I am trying to do is let a function call a controller with some params (if needed) in order to execute some other function dynamically. But it tells me its undefined or a promise that is returned. (I have axios multiple times before and i'm quite oke. The functionalities work on their in in the example, but im trying to make them more dynamic)

Example:
// some function
import API
Todo = () => {
   API.Todo.getList().then((result) => { console.log(result) 
}

// API
const Todo = {
 getList: () => {
   return call.get(id, sort);
 },
}

const call = {
    get: () => {
       1: return axios(‘url_Todo_List’);
                     OR
       2: axios(‘url_Todo_List’).then(response => return response);
    },
}

I somehow try to run something like this, however, it still gives me an error of undefined, thus, then() cannot be executed in the //some function.

What happened during my trials, is that the 1: option in the axios call did work like this for getList(). But not for the // some function .... The question that remains is: “How do I do this when it’s nested like this and is there a way to do the 2nd option which will give me the option to return an error as well?”

I know there is no issue with axios. But, I wish to learn a little but more about axios and its promises in order to learn more.

Thank you in advance :)

**Comment by sbimochan** [comment]
You need to add a REST verb first like GET, POST.
For example:
`axios.get('someUrl).then(function(response) =>console.log(response.data));`

**Comment by D19cannon** [comment]
Thanks for the quick response!
Sorry that's a typo of mine. It's suppose to be a quick example in relation with other functions by using the promise .then() flow.

Any suggestions why the .then() continuous as undefined or as a promise?

**Comment by D19cannon** [comment]
Hi, my question isn't about fixing my code. The code is a just a quick sample. The point of my question is: "How does the promise flows through multiple questions." Since, my problem has more to do with what my promises return such as < undefined / promise >

Again. Don't look at the syntax. I honestly, do not care about that. It's the theoretical part how axios resolves its promises when it's asked through a middle/helper function. 

In other words, three functions that are all promise based, resolve before the first function returns anything.

https://medium.com/@chung.andrew7/a-surprising-functional-programming-utility-of-pending-promises-with-async-await-256d523533b9

**Comment by wasnlosdu** [comment]
So what you mean is _when does the JavaScript finally resolve the Promise?_

**Comment by D19cannon** [comment]
Exactly. Whenever I just run a function such as:
// this works just fine for example
getDetails = () => {
  axios.get('url')
      .then((resp) => {
         let x =  dosomething(resp);
       ......
      })
}

But whenever there are two functions involved and both have a promise it seems that the function that is calling the other resolves in undefined or a promise, before the other function even has a chance to resolve and return it.

------------------------------------------------------------------------------------------------

Which bring me to this solution:
// option 1 works, but two doesn't
getDetails = () => {
  details.get().then( result => { doSomething(result)});
}

const details = {
1:    get: () => {
            return  axios.get('url')
    }

2:   get: () => {
          axios.get('url')
            .then(resp => {
               let x = something(resp);
               return x;
             }
        }
} // end details

You would say know just use the 1 option and problem solved. But what if the flow goes through 3 functions which are all promises?

------------------------------------------------------------------------------------------------

// this doesn't work && group just to get the idea for another function
// request.post() and request.get() are actual function how they should be used
getDetailsList = () => {
  details.list().then( result => { doSomething(result)});
}

const details = {
    list: () => {
            request.get().then(list => { return doSomething(list) }
    }

   group: () => {
      request.get().then(groupData => { return doSomething(groupData)
   }
} // end details

const request = {
   get: () => {
          axios.get('url')
            .then(resp => {
               let x = something(resp);
               return x;
             }
        }

   post: () => {
         return axios({
              method: 'post',
              url: `${API_ROOT}${url}`,
              headers: {
                  'Content-Type': 'application/json',
                  'Accept': 'application/json',
               },
               data: {
                  product_ids: ids,
                  text: filter
               },
              'withCredentials': true
        })
} // end request

This is pretty much the idea of the flow, controlled and reusable axios calls based on what ever the function need such as details().

The first them I can resolve it with is by using: return axios.get('url') but the .then() can't be used in any of the cases. Also, it only works in the 2nd function (details()) to get the response. The original function that is calling it ain't getting the response, because it's a promise or undefined.

I've also tried to do this with async / await with no luck.

**Comment by wasnlosdu** [comment]
So I managed to something like this: https://stackblitz.com/edit/js-p58qhg

A then is used for promises, once a promise has been resolved (e.g. data has been fetched) it can be anything (including another promise). So when you are trying to chain `then` on a `then` it throws an error - because you try to use a `then` on an object not on a promise.

Is this somehow understandable ? :)

**Comment by D19cannon** [comment]
Thanks @wasnlosdu  !

I happen to solve it just before you replied. It was a bit tricky, but it works now :)

I did the following (tried to add it in your stackblitz, but didnt work :(( )
//----------------------------------------------------------------- //

Remember these is sort of an example, obviously the functions will not be executed like this.

The trick was to return axios without .then() and async / await in the controller (otherwise for some reason it could return a promise instead. Also, .then() doesn't work in there)

//----------------------------------------------------------------- //
// ----------------------- React/Mobx component ----------------------------------- //

login = () => {
  request.Auth.login()
  .then( response => {
    doSometing(response);
  })
}

// ----------------------- API calls -- Axios---------------------------- //

import axios from 'axios';

const request = {
  post: ({urlRequest, token = null, username =  '', password = '' }) => {
    return axios({
      method: 'post',
      url: urlRequest,
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'X-CSRF-Token': token,
      },
      data: {
        "username": username,
        "password": password,
      },
      'withCredentials': true
    });
  }
};

// *  All the requests for user authentication and authorization
const Auth = {
  login: async (username, password) => {
    return await request.post(`${API_ROOT}/user/login.json`, username, password);
  }
};
